fixes #2903 accumulated delay and cron jobs alignement

toni-moreno opened the pull request:
Hi With this PR we can fix the behaviour described in #2903.

**With current drone version**

![image](https://user-images.githubusercontent.com/5883405/72604755-d2b1f680-391b-11ea-8e91-acd28202f44b.png)

**With this patch**

![image](https://user-images.githubusercontent.com/5883405/72604826-eeb59800-391b-11ea-867a-c6fe29119dc9.png)

toni-moreno commented on handler/api/repos/crons/create_test.go:
```diff
@@ -56,7 +56,7 @@ func TestHandleCreate(t *testing.T) {
 	json.NewDecoder(w.Body).Decode(got)
 
-	ignore := cmpopts.IgnoreFields(core.Cron{}, "Next")
+	ignore := cmpopts.IgnoreFields(core.Cron{}, "Next", "Prev")
```
Now Prev Field will change on each cron job creation

toni-moreno commented on trigger/cron/cron_test.go:
```diff
@@ -44,9 +44,6 @@ func TestCron(t *testing.T) {
 	checkCron := func(_ context.Context, cron *core.Cron) {
-		if got, want := cron.Prev, int64(2000000000); got != want {
-			t.Errorf("Expect Next copied to Prev")
-		}
```
Prev won't be anymore copied from Next to avoid alignment

tboerger commented on trigger/cron/cron.go:
```diff
@@ -46,12 +46,21 @@ type Scheduler struct {
+			start := time.Now()
 			s.run(ctx)
+			elapsed := time.Since(start)
+			logrus.Infof("cron: job processed in [%s]", elapsed)
```
You could use the structured logger with some fields

tboerger commented on trigger/cron/cron.go:
```diff
@@ -91,17 +100,33 @@ func (s *Scheduler) run(ctx context.Context) error {
+			next = sched.Next(next)
+			// if next still under now this scheduled  will be skipped , else will be executed !
+			if next.Before(now) {
+				logger.Warnf("skipped cron job [%s]  scheduled in the past at : [%s]  NOW [%s]", job.Name, prev, now)
```
Too much spaces?

toni-moreno replied:
Hi @tboerger thank you for the quick reply. 
What exactly would you refer? Are you talking aboult spaces in the log message?

tboerger replied:
Yes, just refering to the duplicated spaces within the log message, so really nothing critical.

tboerger commented on trigger/cron/cron.go:
```diff
@@ -91,17 +100,33 @@ func (s *Scheduler) run(ctx context.Context) error {
+			}
+		}
+
+		logger.Infof("cron job [%s]   scheduled  at : [%s] has been triggered at [%s] ", job.Name, prev, now)
```
Same here

toni-moreno commented:
Hi @tboerger I've rebased the PR and also added a few fixes. 
a) skip first execution , f not skipped all old scheduled jobs will be triggered at once first time , after first execution each one will be triggered with its own timing, not aligned.
b)  when next scheduled time is equal to now(), I was scheduled twice.

Now it works much better than previous PR

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/harness/drone?pullRequest=2904) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/harness/drone?pullRequest=2904) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/harness/drone?pullRequest=2904) it.</sub>

magicoder10 commented:
Can this PR be merge?! Cron has been repeatedly triggering multiple builds simultaneously for me, which wasted lots of my time to add logs and debug Drone's source code. Also wasted me hours trying to migrate from Drone CI to Github actions

bot2-harness commented:
This PR has been automatically closed due to inactivity.
